2_process/src/lake_sequences_mntoha.py

```diff
@@ -0,0 +1,241 @@
+                                    min_days=sequence_length)
+        # obs_full now has dimensions of len(date range) by len(depths)
+
+        # Next, add features (drivers, clarity, ice flags, and static
+        # attributes) to obs_full. There are some NaN values in the clarity and
+        # ice flag data. There are multiple ways to deal with NaN values in
+        # inputs: interpolate to fill them, skip over them, or remove sequences
+        # containing NaNs entirely. At this point, allow NaN values into the
+        # sites. Later code will handle NaN values.
+
+        # Join drivers by date
+        drivers = pd.read_csv(drivers_file, parse_dates=['time'], index_col='time')
+        obs_full = obs_full.join(drivers, how='left')
+
+        # Join clarity by date
+        clarity = pd.read_csv(clarity_file, parse_dates=['date'], index_col='date')
+        obs_full = obs_full.join(clarity, how='left')
+
+        # Join ice flags by date
+        ice_flags = pd.read_csv(ice_flags_file, parse_dates=['date'], index_col='date')
+        obs_full = obs_full.join(ice_flags, how='left')
+
+        # Get attributes
+        obs_full['area'] = lake['area']
+        obs_full['lon'] = lake['centroid_lon']
+        obs_full['lat'] = lake['centroid_lat']
+        obs_full['elevation'] = lake['elevation']
+        # Now obs_full is one long timeseries with depth-specific temperatures,
+        # drivers, clarity, ice_flags, and attributes as columns
+
+        # Split into sequences
+
+        # Let's make some rules for the sequences to save to disk:
+        # 1. Sequences are sequence_length days long
+        # 2. Each next sequence starts sequence_offset days later than the previous
+        # 3. Discard sequences that lack observed temperatures
+        # 4. We know going in that the first and last sequences contain observed temperatures
+
+        # Convert dataframe to numpy array
+        # Use float32 to save on memory and train the neural network more efficiently.
+        obs_full_array = obs_full.to_numpy(dtype=np.float32)
+        # The shape of obs_full_array is (len(date range), # depths + # features)
+        # The first `len(depths)` elements in the second dimension are temperature observations.
+        # Features are drivers, clarity, ice flags, and static attributes, in that order.
+
+        # Create a strided view into obs_full_array
+        # This creates a numpy object with sequences of length sequence_length
+        # and offsets of sequence_offset, without making an in-memory copy of the array
+        all_sequences = np.lib.stride_tricks.sliding_window_view(
+            obs_full_array, sequence_length, axis=0
+        )[::sequence_offset]
+        # The shape of all_sequences is (# sequences, # depths + # features, sequence_length).
+
+        # Only write sequences to file if they contain temperature observations
+        sequences_have_no_obs = np.all(np.isnan(all_sequences[:, :len(depths), :]), axis=(1, 2))
```

Comment:
might just note again here that "# The first `len(depths)` elements in the second dimension are temperature observations."

Reply:
Done!